### PR TITLE
add chat realtime with websocket

### PR DESCRIPTION
**Pull request by ledat139**
# ✅ Checklist Tự review
## 🧹 Code sạch & format chuẩn
- [x] Code sạch, dễ đọc, theo convention của team (naming, cấu trúc)
- [ ] Đã chạy `go fmt`, `goimports`, `golangci-lint` (nếu có)
- [x] Không còn đoạn code tạm, `fmt.Println`, `log.Println`, comment debug
- [x] Không còn biến, hàm, import không sử dụng
- [x] Hàm không quá dài, có thể tách nhỏ nếu vượt quá 50 dòng
- [x] Code không lặp lại, đã tái sử dụng hợp lý

## 🧠 Logic & Xử lý lỗi
- [x] Logic đúng yêu cầu, không dư thừa, rõ ràng
- [x] Xử lý đầy đủ các trường hợp `nil`, `error`, `invalid input`
- [x] Mỗi thao tác có thể lỗi (I/O, DB, parse...) đều được check `err`
- [x] Không để `err` bị silent — luôn log hoặc trả về (tùy ngữ cảnh)
- [x] Sử dụng `context` đúng cách (nếu dùng), tránh leak
- [x] Nếu là `goroutine`, có quản lý lifecycle và tránh leak

## 🌐 HTTP Handler (Gin)
- [x] Sử dụng đúng HTTP method (GET, POST, PUT, DELETE…)
- [x] Validate request (param, query, body) đầy đủ, rõ ràng
- [x] Sử dụng binding (`ShouldBindJSON`, `ShouldBindUri`, v.v.) đúng cách
- [x] Xử lý trả về `status code` phù hợp (`200`, `201`, `400`, `404`, `500`…)
- [x] Response rõ ràng, có `message`, `data`, `error` (nếu cần)
- [x] Không leak thông tin nhạy cảm trong response lỗi

## 🗃️ Struct & Layer
- [x] Struct có annotation JSON đầy đủ (`json:"field_name"`)
- [x] Sử dụng DTO/input/output rõ ràng, không lẫn lộn model DB và response
- [x] Đã tách rõ các layer: handler → service → repository (nếu theo clean code)
- [x] Middleware viết rõ ràng, dễ đọc, không làm chậm hệ thống

## 🧪 Unit Test & Mock
- [x] Có test cho handler/service chính (ít nhất happy case và edge case)
- [x] Nếu dùng DB/service ngoài, có mock hoặc test isolation
- [x] Có test validate đầu vào (request, param, query)
- [x] Không dùng test hardcode giá trị không ổn định (vd: time.Now() mà không mock)
- [x] Tên test rõ ràng: `TestFunction_Scenario_Expected`

## 🔐 Security & Config
- [x] Không hardcode secret, token, config trong code
- [x] Không để lộ thông tin nhạy cảm qua log hoặc panic
- [x] Validate và sanitize đầu vào người dùng
- [x] Nếu có auth, kiểm tra middleware hoạt động đúng

## 🤖 Sử dụng GitHub Copilot
- [x] Đã kiểm tra kỹ mọi đoạn code được gợi ý từ GitHub Copilot
- [ ] Không để lại phần comment hoặc TODO do Copilot sinh ra
- [ ] Không commit đoạn code mà chưa hiểu rõ logic hoặc mục đích
- [ ] Đảm bảo mọi function/code block từ Copilot phù hợp với context dự án
- [ ] Tối ưu lại code được gợi ý nếu cần — tránh lạm dụng hoặc copy nguyên mẫu

## Related Tickets
- https://edu-redmine.sun-asterisk.vn/issues/90478

## WHAT (optional)
- Implement chat-service

## HOW
- Use WebSocket to handle real-time chat between users

## WHY (optional)
- 

## Notes (Kiến thức tìm hiểu thêm)
- websocket in go

### CODE REVIEW

**Review by truongtamthanh2004 — APPROVED** [review]

**Comment by ledat139** [comment]
READY

File: services/chat-service/internal/handler/chat_handler.go [hunk 1]
```diff
@@ -1 +1,53 @@
...
+	if err := h.hub.HandleWebSocket(c.Writer, c.Request); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{
+			"message": err.Error(),
+		})
+		return
+	}
+}
+
+func (h *ChatHandler) GetConversation(c *gin.Context) {
+	user1, err := strconv.Atoi(c.Param("user1"))
```
Comment (chienpv-3590) [inline services/chat-service/internal/handler/chat_handler.go:32]: Nên kiểm tra thêm các trường hợp đặc biệt, ví dụ: hai user trùng nhau, hoặc user không tồn tại.

File: services/chat-service/internal/handler/chat_handler.go [hunk 2]
```diff
@@ -1 +1,53 @@
 package handler
+
+import (
+	"chat-service/internal/constant"
```
Comment (chienpv-3590) [inline services/chat-service/internal/handler/chat_handler.go:4]: Nếu không sử dụng hết các package (ví dụ: nếu không dùng "chat-service/internal/constant" ngoài lỗi), có thể cân nhắc bỏ đi.

File: services/chat-service/internal/repository/chat_repository.go [hunk 3]
```diff
@@ -1 +1,34 @@
...
+func NewChatRepository(db *gorm.DB) ChatRepository {
+	return &chatRepository{db: db}
+}
+
+func (r *chatRepository) SaveMessage(ctx context.Context, msg *model.ChatMessage) error {
+	return r.db.WithContext(ctx).Create(msg).Error
+}
+
+func (r *chatRepository) GetConversation(ctx context.Context, user1, user2 uint) ([]model.ChatMessage, error) {
+	var messages []model.ChatMessage
```
Comment (chienpv-3590) [inline services/chat-service/internal/repository/chat_repository.go:28]: Nếu user không tồn tại, vẫn thực hiện query và trả về empty array. Có thể bổ sung kiểm tra đầu vào nếu cần bảo mật hơn.

**Comment by ledat139** [comment]
READY
